===============================================================================

This is a patch for the Netkit user-mode Linux kernel.
For more information: http://www.netkit.org/

This patch instructs the UML kernel to look for the port-helper tool inside
Netkit directories first.
Virtual machines listening for telnet connections spawn a port-helper every time
a new connection is established.

===============================================================================

     Copyright (C) 2015  Massimo Rimondini (rimondini@ing.uniroma3.it)
     Computer Networks Research Group, Roma Tre University
 
     This program is free software: you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation, version 3 of the License.
 
     This program is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     GNU General Public License for more details.
 
     You should have received a copy of the GNU General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

=================================================================================
diff -Naur linux-3.19.3-original/arch/um/drivers/port_user.c linux-3.19.3/arch/um/drivers/port_user.c
--- linux-3.19.3-original/arch/um/drivers/port_user.c	2015-03-26 14:00:21.000000000 +0100
+++ linux-3.19.3/arch/um/drivers/port_user.c	2015-04-03 10:46:20.000000000 +0200
@@ -5,6 +5,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <errno.h>
 #include <termios.h>
 #include <unistd.h>
@@ -167,9 +168,24 @@
 int port_connection(int fd, int *socket, int *pid_out)
 {
 	int new, err;
-	char *argv[] = { "/usr/sbin/in.telnetd", "-L",
-			 "/usr/lib/uml/port-helper", NULL };
+	/************************************************/
+	/* Changed for better compatibility with Netkit */
+	/************************************************/
+	/* char *argv[] = { "/usr/sbin/in.telnetd", "-L",
+			 "/usr/lib/uml/port-helper", NULL }; */
 	struct port_pre_exec_data data;
+	char *netkit_home, port_helper_bin[4096];
+
+	netkit_home = getenv("NETKIT_HOME");
+	if (!netkit_home)
+		strcpy (port_helper_bin, "/usr/lib/uml/port-helper");
+	else
+		sprintf (port_helper_bin, "%s/bin/port-helper", netkit_home);
+
+	char *argv[] = { "/usr/sbin/in.telnetd", "-L",
+			 port_helper_bin, NULL };
+
+	/*****************************************/
 
 	new = accept(fd, NULL, 0);
 	if (new < 0)
